kernel/dynamic.c

```diff
@@ -120,20 +121,14 @@ static void dyn_cb(const struct k_thread *thread, void *user_data)
 
 int z_impl_k_thread_stack_free(k_thread_stack_t *stack)
 {
-	char state_buf[16] = {0};
 	struct dyn_cb_data data = {.stack = stack};
 
 	/* Get a possible tid associated with stack */
 	k_thread_foreach(dyn_cb, &data);
 
 	if (data.tid != NULL) {
-		/* Check if thread is in use */
-		if (k_thread_state_str(data.tid, state_buf, sizeof(state_buf)) != state_buf) {
-			LOG_ERR("tid %p is invalid!", data.tid);
-			return -EINVAL;
-		}
-
-		if (!(strcmp("dummy", state_buf) == 0) || (strcmp("dead", state_buf) == 0)) {
+		if (!(z_is_thread_state_set(data.tid, _THREAD_DUMMY) ||
```

Comment:
@finikorg - if we wanted to avoid race conditions and locking requirements, we could improve the `z_is_thread_state_set()` call to accept a bitmask of possible thread states, returning true if any of the specified states are set.